File: matrix.java
import java.util.*;

public class matrix {

    public static int[][] inputMatrix(Scanner sc) {
        System.out.print("Masukkan jumlah baris: ");
        int rows = sc.nextInt();

        System.out.print("Masukkan jumlah kolom: ");
        int cols = sc.nextInt();

        int[][] matrix = new int[rows][cols];

        System.out.println("Masukkan elemen matrix:");
        for (int i = 0; i < rows; i++) {
            while (true) {
                System.out.print("Baris " + (i + 1) + ": ");

                String line = sc.nextLine();
                if (line.isEmpty()) {
                    line = sc.nextLine();
                }

                String[] input = line.split(" ");

                if (input.length != cols) {
                    System.out.println("Jumlah elemen harus " + cols);
                    continue;
                }

                for (int j = 0; j < cols; j++) {
                    matrix[i][j] = Integer.parseInt(input[j]);
                }
                break;
            }
        }

        return matrix;
    }

    public static void printMatrix(int[][] matrix) {
        System.out.println("\nMatrix:");
        for (int[] row : matrix) {
            for (int value : row) {
                System.out.print(value + " ");
            }
            System.out.println();
        }
        System.out.println();
    }

    public static int[][] sortRowWise(int[][] matrix) {
        int[][] result = new int[matrix.length][matrix[0].length];

        for (int i = 0; i < matrix.length; i++) {
            result[i] = matrix[i].clone();
            Arrays.sort(result[i]);
        }

        return result;
    }

    public static int[][] sortColumnWise(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            result[i] = matrix[i].clone();
        }

        for (int j = 0; j < cols; j++) {
            int[] col = new int[rows];

            for (int i = 0; i < rows; i++) {
                col[i] = matrix[i][j];
            }

            Arrays.sort(col);

            for (int i = 0; i < rows; i++) {
                result[i][j] = col[i];
            }
        }

        return result;
    }

    public static int[][] rotateClockwiseByOne(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        if (rows < 2 || cols < 2) {
            return matrix;
        }

        int[][] result = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            result[i] = matrix[i].clone();
        }

        int temp = matrix[1][0];

        for (int i = 0; i < cols; i++) {
            int t = result[0][i];
            result[0][i] = temp;
            temp = t;
        }

        for (int i = 1; i < rows; i++) {
            int t = result[i][cols - 1];
            result[i][cols - 1] = temp;
            temp = t;
        }

        for (int i = cols - 2; i >= 0; i--) {
            int t = result[rows - 1][i];
            result[rows - 1][i] = temp;
            temp = t;
        }

        for (int i = rows - 2; i >= 0; i--) {
            int t = result[i][0];
            result[i][0] = temp;
            temp = t;
        }

        return result;
    }

    public static int[][] rotateCounterClockwiseByOne(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        if (rows < 2 || cols < 2) {
            return matrix;
        }

        int[][] result = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            result[i] = matrix[i].clone();
        }

        int temp = matrix[0][1];

        for (int i = 0; i < rows; i++) {
            int t = result[i][0];
            result[i][0] = temp;
            temp = t;
        }

        for (int i = 1; i < cols; i++) {
            int t = result[rows - 1][i];
            result[rows - 1][i] = temp;
            temp = t;
        }

        for (int i = rows - 2; i >= 0; i--) {
            int t = result[i][cols - 1];
            result[i][cols - 1] = temp;
            temp = t;
        }

        for (int i = cols - 2; i >= 0; i--) {
            int t = result[0][i];
            result[0][i] = temp;
            temp = t;
        }

        return result;
    }

    public static int[][] rotate90(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][rows - 1 - i] = matrix[i][j];
            }
        }

        return result;
    }

    public static int[][] rotate180(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[rows - 1 - i][cols - 1 - j] = matrix[i][j];
            }
        }

        return result;
    }

    public static void rowWiseTraversal(int[][] matrix) {
        System.out.println("Row-wise Traversal:");
        for (int[] row : matrix) {
            for (int item : row) {
                System.out.print(item + " ");
            }
        }
        System.out.println();
    }

    public static void columnWiseTraversal(int[][] matrix) {
        System.out.println("Column-wise Traversal:");

        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int j = 0; j < cols; j++) {
            for (int i = 0; i < rows; i++) {
                System.out.print(matrix[i][j] + " ");
            }
        }

        System.out.println();
    }

    public static void spiralPrint(int[][] matrix) {
        System.out.println("Spiral Traversal:");

        int top = 0;
        int bottom = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;

        while (top <= bottom && left <= right) {

            for (int i = left; i <= right; i++) {
                System.out.print(matrix[top][i] + " ");
            }
            top++;

            for (int i = top; i <= bottom; i++) {
                System.out.print(matrix[i][right] + " ");
            }
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    System.out.print(matrix[bottom][i] + " ");
                }
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    System.out.print(matrix[i][left] + " ");
                }
                left++;
            }
        }

        System.out.println();
    }

    public static int[][] transpose(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][i] = matrix[i][j];
            }
        }

        return result;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int[][] matrix = inputMatrix(sc);

        while (true) {
            System.out.println("\n========== MENU ==========");
            System.out.println("1-a. Sort the matrix row-wise");
            System.out.println("1-b. Sort the matrix column-wise");
            System.out.println("2-a. Rotate Matrix Clockwise by 1");
            System.out.println("2-b. Rotate Matrix Counter-Clockwise by 1");
            System.out.println("2-c. Rotate a matrix by 90");
            System.out.println("2-d. Rotate a matrix by 180");
            System.out.println("3-a. Row-wise traversal of matrix");
            System.out.println("3-b. Column-wise traversal of matrix");
            System.out.println("4. Print matrix in spiral form");
            System.out.println("5. Transpose matrix");
            System.out.println("6. Quit");

            System.out.print("\nPilih menu: ");
            String choice = sc.next().toLowerCase();

            if (choice.equals("1-a")) {
                printMatrix(sortRowWise(matrix));

            } else if (choice.equals("1-b")) {
                printMatrix(sortColumnWise(matrix));

            } else if (choice.equals("2-a")) {
                printMatrix(rotateClockwiseByOne(matrix));

            } else if (choice.equals("2-b")) {
                printMatrix(rotateCounterClockwiseByOne(matrix));

            } else if (choice.equals("2-c")) {
                printMatrix(rotate90(matrix));

            } else if (choice.equals("2-d")) {
                printMatrix(rotate180(matrix));

            } else if (choice.equals("3-a")) {
                rowWiseTraversal(matrix);

            } else if (choice.equals("3-b")) {
                columnWiseTraversal(matrix);

            } else if (choice.equals("4")) {
                spiralPrint(matrix);

            } else if (choice.equals("5")) {
                printMatrix(transpose(matrix));

            } else if (choice.equals("6")) {
                System.out.println("Program selesai.");
                break;

            } else {
                System.out.println("Pilihan tidak valid!");
            }
        }

        sc.close();
    }
}

File: matrix.py
def input_matrix():
    rows = int(input("Masukkan jumlah baris: "))
    cols = int(input("Masukkan jumlah kolom: "))

    matrix = []
    print("Masukkan elemen matrix:")
    for i in range(rows):
        row = list(map(int, input(f"Baris {i+1}: ").split()))
        while len(row) != cols:
            print(f"Jumlah elemen harus {cols}")
            row = list(map(int, input(f"Baris {i+1}: ").split()))
        matrix.append(row)

    return matrix


def print_matrix(matrix):
    print("\nMatrix:")
    for row in matrix:
        print(*row)
    print()


def sort_row_wise(matrix):
    return [sorted(row) for row in matrix]


def sort_column_wise(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    result = [row[:] for row in matrix]

    for j in range(cols):
        col = [matrix[i][j] for i in range(rows)]
        col.sort()

        for i in range(rows):
            result[i][j] = col[i]

    return result


def rotate_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    if rows < 2 or cols < 2:
        return matrix

    result = [row[:] for row in matrix]

    temp = matrix[1][0]

    for i in range(cols):
        result[0][i], temp = temp, result[0][i]

    for i in range(1, rows):
        result[i][cols-1], temp = temp, result[i][cols-1]

    for i in range(cols-2, -1, -1):
        result[rows-1][i], temp = temp, result[rows-1][i]

    for i in range(rows-2, -1, -1):
        result[i][0], temp = temp, result[i][0]

    return result


def rotate_counter_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    if rows < 2 or cols < 2:
        return matrix

    result = [row[:] for row in matrix]

    temp = matrix[0][1]

    for i in range(rows):
        result[i][0], temp = temp, result[i][0]

    for i in range(1, cols):
        result[rows-1][i], temp = temp, result[rows-1][i]


    for i in range(rows-2, -1, -1):
        result[i][cols-1], temp = temp, result[i][cols-1]

    for i in range(cols-2, -1, -1):
        result[0][i], temp = temp, result[0][i]

    return result


def rotate_90(matrix):
    return [list(row) for row in zip(*matrix[::-1])]


def rotate_180(matrix):
    return [row[::-1] for row in matrix[::-1]]


def row_wise_traversal(matrix):
    print("Row-wise Traversal:")
    for row in matrix:
        for item in row:
            print(item, end=" ")
    print()


def column_wise_traversal(matrix):
    print("Column-wise Traversal:")
    rows = len(matrix)
    cols = len(matrix[0])

    for j in range(cols):
        for i in range(rows):
            print(matrix[i][j], end=" ")
    print()


def spiral_print(matrix):
    print("Spiral Traversal:")

    top = 0
    bottom = len(matrix) - 1
    left = 0
    right = len(matrix[0]) - 1

    while top <= bottom and left <= right:

        for i in range(left, right + 1):
            print(matrix[top][i], end=" ")
        top += 1

        for i in range(top, bottom + 1):
            print(matrix[i][right], end=" ")
        right -= 1

        if top <= bottom:
            for i in range(right, left - 1, -1):
                print(matrix[bottom][i], end=" ")
            bottom -= 1

        if left <= right:
            for i in range(bottom, top - 1, -1):
                print(matrix[i][left], end=" ")
            left += 1

    print()


def transpose(matrix):
    return [list(row) for row in zip(*matrix)]


matrix = input_matrix()

while True:
    print("\n========== MENU ==========")
    print("1-a. Sort the matrix row-wise")
    print("1-b. Sort the matrix column-wise")
    print("2-a. Rotate Matrix Clockwise by 1")
    print("2-b. Rotate Matrix Counter-Clockwise by 1")
    print("2-c. Rotate a matrix by 90")
    print("2-d. Rotate a matrix by 180")
    print("3-a. Row-wise traversal of matrix")
    print("3-b. Column-wise traversal of matrix")
    print("4. Print matrix in spiral form")
    print("5. Transpose matrix")
    print("6. Quit")

    choice = input("\nPilih menu: ").lower()

    if choice == "1-a":
        print_matrix(sort_row_wise(matrix))

    elif choice == "1-b":
        print_matrix(sort_column_wise(matrix))

    elif choice == "2-a":
        print_matrix(rotate_clockwise_by_one(matrix))

    elif choice == "2-b":
        print_matrix(rotate_counter_clockwise_by_one(matrix))

    elif choice == "2-c":
        print_matrix(rotate_90(matrix))

    elif choice == "2-d":
        print_matrix(rotate_180(matrix))

    elif choice == "3-a":
        row_wise_traversal(matrix)

    elif choice == "3-b":
        column_wise_traversal(matrix)

    elif choice == "4":
        spiral_print(matrix)

    elif choice == "5":
        print_matrix(transpose(matrix))

    elif choice == "6":
        print("Program selesai.")
        break

    else:
        print("Pilihan tidak valid!")